src/mars.c

```diff
@@ -1694,21 +1710,74 @@ Language changes listed by -transition=id:\n\
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
 
             obj_start(m->srcfile->toChars());
-            genObjFile(m, global.params.multiobj);
+            genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
-                genObjFile(entrypoint, global.params.multiobj);
+                genObjFile(entrypoint, false);
             for (size_t j = 0; j < Module::amodules.dim; j++)
             {
+                // todo: This part is not the best for the total size of object files.
+                // For example:
+                // When both of main.d and code.d import mx, even if main.obj
+                // already contains mx->marray, it is stored in code.obj again.
                 Module *mx = Module::amodules[j];
-                if (mx != m && mx->importedFrom == m && (mx->marray || mx->massert || mx->munittest))
-                    genhelpers(mx, true);
+                if (mx == m || mx->importedFrom != m)
+                    continue;
+                if (!mx->marray && !mx->massert && !mx->munittest)
+                    continue;
+                genhelpers(mx, true);
```

Comment:
Still emitting weak duplicates of those helper functions into other modules, means it won't solve [Issue 14748 – Removing std.stdio import causes 2x increase in "Hello, world" program binary filesize](https://issues.dlang.org/show_bug.cgi?id=14748) for separate compilation.
This indicates a severe flaw of this approach.